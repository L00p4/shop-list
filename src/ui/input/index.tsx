import React, { forwardRef } from 'react'

import { WrapperInput, StyledInput, Label, ErrorMessage } from './input.styles'

export { default as InputSkeleton } from './input.skeleton'

export type InputProps = {
  label?: string
  error?: string
  variant?: 'default' | 'filled'
  fullWidth?: boolean
} & React.InputHTMLAttributes<HTMLInputElement>

const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      label,
      error,
      variant = 'default',
      fullWidth = false,
      className,
      id,
      ...rest
    },
    ref
  ) => {
    const inputId = id || label?.toLowerCase().replace(/\s+/g, '-')

    return (
      <WrapperInput fullWidth={fullWidth} className={className}>
        {label && <Label htmlFor={inputId}>{label}</Label>}
        <StyledInput
          ref={ref}
          id={inputId}
          variant={variant}
          hasError={!!error}
          {...rest}
        />
        {error && <ErrorMessage>{error}</ErrorMessage>}
      </WrapperInput>
    )
  }
)

Input.displayName = 'Input'

export default Input

// Input/index.tsx
// import React, { forwardRef, useId } from 'react'
// import {
//   WrapperInput,
//   StyledInput,
//   Label,
//   ErrorMessage,
//   HintText
// } from './input.styles'

// export { default as InputSkeleton } from './input.skeleton'

// export interface InputProps
//   extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
//   label?: string
//   error?: string
//   hint?: string
//   variant?: 'default' | 'filled'
//   size?: 'small' | 'medium' | 'large'
//   fullWidth?: boolean
//   leftIcon?: React.ReactNode
//   rightIcon?: React.ReactNode
// }

// const Input = forwardRef<HTMLInputElement, InputProps>(
//   (
//     {
//       label,
//       error,
//       hint,
//       variant = 'default',
//       size = 'medium',
//       fullWidth = false,
//       leftIcon,
//       rightIcon,
//       className,
//       id,
//       disabled,
//       ...rest
//     },
//     ref
//   ) => {
//     // Hook seguro para gerar ID Ãºnico
//     const generatedId = useId()
//     const inputId = id || `input-${generatedId}`
//     const errorId = error ? `${inputId}-error` : undefined
//     const hintId = hint ? `${inputId}-hint` : undefined

//     return (
//       <WrapperInput
//         fullWidth={fullWidth}
//         hasError={!!error}
//         className={className}
//       >
//         {label && (
//           <Label htmlFor={inputId} disabled={disabled}>
//             {label}
//           </Label>
//         )}

//         <div className="input-container">
//           {leftIcon && (
//             <span
//               className="input-icon input-icon--left"
//               data-testid="left-icon"
//             >
//               {leftIcon}
//             </span>
//           )}

//           <StyledInput
//             ref={ref}
//             id={inputId}
//             variant={variant}
//             size={size}
//             hasError={!!error}
//             hasLeftIcon={!!leftIcon}
//             hasRightIcon={!!rightIcon}
//             disabled={disabled}
//             aria-invalid={!!error}
//             aria-describedby={
//               [errorId, hintId].filter(Boolean).join(' ') || undefined
//             }
//             {...rest}
//           />

//           {rightIcon && (
//             <span
//               className="input-icon input-icon--right"
//               data-testid="right-icon"
//             >
//               {rightIcon}
//             </span>
//           )}
//         </div>

//         {hint && !error && <HintText id={hintId}>{hint}</HintText>}

//         {error && (
//           <ErrorMessage id={errorId} role="alert">
//             {error}
//           </ErrorMessage>
//         )}
//       </WrapperInput>
//     )
//   }
// )

// Input.displayName = 'Input'

// export default Input
