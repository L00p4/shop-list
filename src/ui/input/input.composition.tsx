import Input, { InputSkeleton } from '.'

export const Primary = () => <Input />

export const Skeleton = () => <InputSkeleton />
