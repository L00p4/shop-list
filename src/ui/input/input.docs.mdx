---
description: Descrição do componente
labels: ['componente']
---
