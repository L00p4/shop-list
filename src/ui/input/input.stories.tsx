// import React from 'react'

// import { Meta, StoryObj } from '@storybook/react'

// import Input, { InputSkeleton } from '.'

// const meta: Meta<typeof Input> = {
//   title: 'Input',
//   component: Input,
//   tags: ['autodocs'],
//   args: {}
// }

// type Story = StoryObj<typeof Input>

// export const Primary: Story = {
//   args: {}
// }

// export const Skeleton: Story = {
//   render: () => <InputSkeleton />,
//   args: {}
// }

// export default meta

// Input/input.composition.tsx
import { Meta, StoryObj } from '@storybook/react'

import Input, { InputSkeleton } from '.'

const meta: Meta<typeof Input> = {
  title: 'UI/Input',
  component: Input,
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: ['default', 'filled']
    },
    type: {
      control: { type: 'select' },
      options: ['text', 'number', 'email', 'password', 'tel']
    },
    fullWidth: {
      control: { type: 'boolean' }
    },
    disabled: {
      control: { type: 'boolean' }
    }
  },
  args: {
    label: 'Label do Input',
    placeholder: 'Placeholder...',
    variant: 'default',
    fullWidth: false,
    disabled: false
  }
}

type Story = StoryObj<typeof Input>

// Basic variants
export const Default: Story = {
  args: {
    label: 'Nome do Produto',
    placeholder: 'Digite o nome do produto...'
  }
}

export const Filled: Story = {
  args: {
    label: 'Nome do Produto',
    placeholder: 'Digite o nome do produto...',
    variant: 'filled'
  }
}

// With error
export const WithError: Story = {
  args: {
    label: 'Preço unitário',
    placeholder: '0,00',
    type: 'number',
    error: 'Preço é obrigatório',
    value: ''
  }
}

// Different input types
export const NumberInput: Story = {
  args: {
    label: 'Quantidade',
    placeholder: '1',
    type: 'number',
    min: 1,
    step: 1
  }
}

export const PriceInput: Story = {
  args: {
    label: 'Preço unitário (R$)',
    placeholder: '0,00',
    type: 'number',
    min: 0,
    step: 0.01
  }
}

// States
export const Disabled: Story = {
  args: {
    label: 'Campo desabilitado',
    placeholder: 'Não é possível editar',
    value: 'Valor fixo',
    disabled: true
  }
}

export const FullWidth: Story = {
  args: {
    label: 'Campo de largura total',
    placeholder: 'Ocupa toda a largura disponível',
    fullWidth: true
  }
}

// Real usage examples from our app
export const ProductNameInput: Story = {
  args: {
    label: 'Nome do Item',
    placeholder: 'Ex: Arroz 5kg, Feijão preto...',
    variant: 'filled',
    fullWidth: true
  },
  parameters: {
    docs: {
      description: {
        story: 'Input para adicionar/editar nome de produto na lista'
      }
    }
  }
}

export const QuantityInput: Story = {
  args: {
    label: 'Quantidade',
    type: 'number',
    min: 1,
    value: '1',
    variant: 'filled'
  },
  parameters: {
    docs: {
      description: {
        story: 'Input para quantidade no modal de adicionar ao carrinho'
      }
    }
  }
}

export const PriceInputModal: Story = {
  args: {
    label: 'Preço unitário (R$)',
    type: 'number',
    min: 0,
    step: 0.01,
    placeholder: '0,00',
    variant: 'filled'
  },
  parameters: {
    docs: {
      description: {
        story: 'Input para preço no modal de adicionar ao carrinho'
      }
    }
  }
}

export const WeightInput: Story = {
  args: {
    label: 'Peso (kg)',
    type: 'number',
    min: 0,
    step: 0.001,
    placeholder: '0,000',
    variant: 'filled'
  },
  parameters: {
    docs: {
      description: {
        story: 'Input para peso de produtos vendidos por quilograma'
      }
    }
  }
}

// Form example
export const FormExample: Story = {
  render: () => (
    <div
      style={{
        display: 'flex',
        flexDirection: 'column',
        gap: 'var(--space-4)',
        maxWidth: '300px'
      }}
    >
      <Input
        label="Nome do Produto"
        placeholder="Ex: Arroz 5kg"
        variant="filled"
        fullWidth
      />
      <div style={{ display: 'flex', gap: 'var(--space-4)' }}>
        <Input
          label="Quantidade"
          type="number"
          min={1}
          value="1"
          variant="filled"
        />
        <Input
          label="Preço (R$)"
          type="number"
          min={0}
          step={0.01}
          placeholder="0,00"
          variant="filled"
        />
      </div>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Exemplo de como os inputs ficam em um formulário real'
      }
    }
  }
}

export const Skeleton: Story = {
  render: () => <InputSkeleton />,
  args: {}
}

export default meta
