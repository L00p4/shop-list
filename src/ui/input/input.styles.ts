import styled, { css } from 'styled-components'

type WrapperInputProps = {
  fullWidth: boolean
}

type StyledInputProps = {
  variant: 'default' | 'filled'
  hasError: boolean
}

const inputVariants = {
  default: css<StyledInputProps>`
    background: transparent;
    border: 1px solid var(--border-default);

    &:focus {
      border-color: var(--purple-500);
      box-shadow: 0 0 0 1px var(--purple-500);
    }

    ${({ hasError }) =>
      hasError &&
      css`
        border-color: var(--error);

        &:focus {
          border-color: var(--error);
          box-shadow: 0 0 0 1px var(--error);
        }
      `}
  `,

  filled: css<StyledInputProps>`
    background: var(--bg-tertiary);
    border: 1px solid transparent;

    &:focus {
      background: var(--bg-secondary);
      border-color: var(--purple-500);
      box-shadow: 0 0 0 1px var(--purple-500);
    }

    &:hover:not(:focus) {
      background: var(--bg-secondary);
    }

    ${({ hasError }) =>
      hasError &&
      css`
        background: rgba(239, 68, 68, 0.1);
        border-color: var(--error);

        &:focus {
          background: rgba(239, 68, 68, 0.1);
          border-color: var(--error);
          box-shadow: 0 0 0 1px var(--error);
        }
      `}
  `
}

export const WrapperInput = styled.div<WrapperInputProps>`
  display: flex;
  flex-direction: column;
  gap: var(--space-2);

  ${({ fullWidth }) =>
    fullWidth &&
    css`
      width: 100%;
    `}
`

export const Label = styled.label`
  color: var(--text-secondary);
  font-size: var(--text-sm);
  font-weight: var(--font-medium);

  /* Cursor pointer to indicate it's clickable */
  cursor: pointer;
`

export const StyledInput = styled.input<StyledInputProps>`
  /* Reset */
  outline: none;
  font-family: inherit;

  /* Layout */
  padding: var(--space-3) var(--space-4);
  border-radius: var(--radius-md);

  /* Typography */
  font-size: var(--text-base);
  color: var(--text-primary);

  /* Transitions */
  transition: all var(--transition-normal);

  /* Apply variant styles */
  ${({ variant }) => inputVariants[variant]}

  /* Placeholder styles */
  &::placeholder {
    color: var(--text-tertiary);
    opacity: 1; /* Firefox fix */
  }

  /* Disabled state */
  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
    background: var(--bg-tertiary);
    color: var(--text-disabled);
  }

  /* Number input specific styles */
  &[type='number'] {
    /* Remove spinner arrows in Chrome, Safari, Edge */
    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button {
      -webkit-appearance: none;
      margin: 0;
    }

    /* Remove spinner arrows in Firefox */
    -moz-appearance: textfield;
  }

  /* Touch improvements for mobile */
  touch-action: manipulation;
  -webkit-tap-highlight-color: transparent;
`

export const ErrorMessage = styled.span`
  color: var(--error);
  font-size: var(--text-xs);
  font-weight: var(--font-medium);
  margin-top: calc(var(--space-2) * -1);
  margin-bottom: var(--space-1);
`

// Input/input.styles.ts
// import styled, { css } from 'styled-components'

// interface WrapperInputProps {
//   fullWidth: boolean
//   hasError?: boolean // Adicionar esta prop
// }

// interface LabelProps {
//   disabled?: boolean
// }

// interface StyledInputProps {
//   variant: 'default' | 'filled'
//   size: 'small' | 'medium' | 'large'
//   hasError: boolean
//   hasLeftIcon: boolean
//   hasRightIcon: boolean
// }

// const inputVariants = {
//   default: css<StyledInputProps>`
//     background: transparent;
//     border: 1px solid var(--border-default);

//     &:focus {
//       border-color: var(--purple-500);
//       box-shadow: 0 0 0 1px var(--purple-500);
//     }

//     ${({ hasError }) =>
//       hasError &&
//       css`
//         border-color: var(--error);

//         &:focus {
//           border-color: var(--error);
//           box-shadow: 0 0 0 1px var(--error);
//         }
//       `}
//   `,

//   filled: css<StyledInputProps>`
//     background: var(--bg-tertiary);
//     border: 1px solid transparent;

//     &:focus {
//       background: var(--bg-secondary);
//       border-color: var(--purple-500);
//       box-shadow: 0 0 0 1px var(--purple-500);
//     }

//     &:hover:not(:focus) {
//       background: var(--bg-secondary);
//     }

//     ${({ hasError }) =>
//       hasError &&
//       css`
//         background: var(--error-bg);
//         border-color: var(--error);

//         &:focus {
//           background: var(--error-bg);
//           border-color: var(--error);
//           box-shadow: 0 0 0 1px var(--error);
//         }
//       `}
//   `
// }

// const inputSizes = {
//   small: css`
//     height: var(--height-input-sm);
//     font-size: var(--text-sm);
//   `,
//   medium: css`
//     height: var(--height-input-md);
//     font-size: var(--text-base);
//   `,
//   large: css`
//     height: var(--height-input-lg);
//     font-size: var(--text-lg);
//   `
// }

// export const WrapperInput = styled.div<WrapperInputProps>`
//   display: flex;
//   flex-direction: column;
//   gap: var(--space-2);

//   ${({ fullWidth }) =>
//     fullWidth &&
//     css`
//       width: 100%;
//     `}

//   /* Input container for icons */
//   .input-container {
//     position: relative;
//     display: flex;
//     align-items: center;
//   }

//   /* Icon styles */
//   .input-icon {
//     position: absolute;
//     display: flex;
//     align-items: center;
//     justify-content: center;
//     color: var(--text-tertiary);
//     z-index: var(--z-base);
//     pointer-events: none;
//     transition: color var(--transition-normal);

//     &--left {
//       left: var(--space-3);
//     }

//     &--right {
//       right: var(--space-3);
//     }
//   }

//   /* Change icon color when input is focused */
//   &:focus-within .input-icon {
//     color: var(--purple-500);
//   }

//   /* Change icon color when there's an error */
//   ${({ hasError }) =>
//     hasError &&
//     css`
//       &:focus-within .input-icon {
//         color: var(--error);
//       }
//     `}
// `

// export const Label = styled.label<LabelProps>`
//   color: var(--text-secondary);
//   font-size: var(--text-sm);
//   font-weight: var(--font-medium);
//   cursor: pointer;

//   ${({ disabled }) =>
//     disabled &&
//     css`
//       color: var(--text-disabled);
//       cursor: not-allowed;
//     `}
// `

// export const StyledInput = styled.input<StyledInputProps>`
//   /* Reset */
//   outline: none;
//   font-family: inherit;
//   width: 100%;

//   /* Layout - padding adjusts based on icons */
//   padding: var(--space-3) var(--space-4);
//   ${({ hasLeftIcon }) =>
//     hasLeftIcon &&
//     css`
//       padding-left: var(--space-10);
//     `}
//   ${({ hasRightIcon }) =>
//     hasRightIcon &&
//     css`
//       padding-right: var(--space-10);
//     `}

//   border-radius: var(--radius-md);

//   /* Typography */
//   color: var(--text-primary);

//   /* Transitions */
//   transition: all var(--transition-normal);

//   /* Apply variant styles */
//   ${({ variant }) => inputVariants[variant]}

//   /* Apply size styles */
//   ${({ size }) => inputSizes[size]}

//   /* Placeholder styles */
//   &::placeholder {
//     color: var(--text-tertiary);
//     opacity: 1; /* Firefox fix */
//   }

//   /* Disabled state */
//   &:disabled {
//     opacity: 0.6;
//     cursor: not-allowed;
//     background: var(--bg-tertiary);
//     color: var(--text-disabled);

//     &::placeholder {
//       color: var(--text-disabled);
//     }
//   }

//   /* Number input specific styles */
//   &[type='number'] {
//     /* Remove spinner arrows in Chrome, Safari, Edge */
//     &::-webkit-outer-spin-button,
//     &::-webkit-inner-spin-button {
//       -webkit-appearance: none;
//       margin: 0;
//     }

//     /* Remove spinner arrows in Firefox */
//     -moz-appearance: textfield;
//   }

//   /* Touch improvements for mobile */
//   touch-action: manipulation;
//   -webkit-tap-highlight-color: transparent;
// `

// export const ErrorMessage = styled.span`
//   color: var(--error);
//   font-size: var(--text-xs);
//   font-weight: var(--font-medium);
// `

// export const HintText = styled.span`
//   color: var(--text-tertiary);
//   font-size: var(--text-xs);
// `
