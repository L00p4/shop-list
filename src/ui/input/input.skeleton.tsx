import { WrapperInput } from './input.styles'

const InputSkeleton = () => (
  <WrapperInput fullWidth>
    <h1 className="skeleton">Input</h1>
  </WrapperInput>
)

export default InputSkeleton
