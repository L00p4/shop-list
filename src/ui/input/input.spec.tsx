import '@testing-library/jest-dom'
import 'jest-environment-jsdom'
import 'jest-styled-components'

import { render } from '@testing-library/react'

import Input from '.'
import { Primary, Skeleton } from './input.composition'

describe('<Input />', () => {
  it('should render the heading', () => {
    render(<Input />)
  })

  it('should render Primary', () => {
    expect(render(<Primary />).container).toMatchSnapshot()
  })

  it('should render Skeleton', () => {
    expect(render(<Skeleton />).container).toMatchSnapshot()
  })
})
