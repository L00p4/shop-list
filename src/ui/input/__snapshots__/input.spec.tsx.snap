// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Input /> should render Primary 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: var(--space-2);
}

.c1 {
  outline: none;
  font-family: inherit;
  padding: var(--space-3) var(--space-4);
  border-radius: var(--radius-md);
  font-size: var(--text-base);
  color: var(--text-primary);
  transition: all var(--transition-normal);
  background: transparent;
  border: 1px solid var(--border-default);
  touch-action: manipulation;
  -webkit-tap-highlight-color: transparent;
}

.c1:focus {
  border-color: var(--purple-500);
  box-shadow: 0 0 0 1px var(--purple-500);
}

.c1::placeholder {
  color: var(--text-tertiary);
  opacity: 1;
}

.c1:disabled {
  opacity: 0.6;
  cursor: not-allowed;
  background: var(--bg-tertiary);
  color: var(--text-disabled);
}

.c1[type='number'] {
  -moz-appearance: textfield;
}

.c1[type='number']::-webkit-outer-spin-button,
.c1[type='number']::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      variant="default"
    />
  </div>
</div>
`;

exports[`<Input /> should render Skeleton 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: var(--space-2);
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <h1
      class="skeleton"
    >
      Input
    </h1>
  </div>
</div>
`;
